src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt

```diff
@@ -171,44 +274,81 @@ private fun StringBuilder.appendDeadlockWithDumpFailure(failure: DeadlockWithDum
-                && "org.jetbrains.kotlinx.lincheck.runner" !in line
-                && "org.jetbrains.kotlinx.lincheck.UtilsKt" !in line
-        }.forEach { appendLine("\t$it") }
+        stackTraceRepresentation(stackTrace).forEach { appendLine("\t$it") }
     }
     return this
 }
 
-private fun StringBuilder.appendIncorrectResultsFailure(failure: IncorrectResultsFailure): StringBuilder {
+private fun StringBuilder.appendIncorrectResultsFailure(
+    failure: IncorrectResultsFailure,
+    exceptionsDescription: Map<Throwable, ExceptionDescription>
+): StringBuilder {
     appendln("= Invalid execution results =")
     if (failure.scenario.initExecution.isNotEmpty()) {
         appendln("Init part:")
-        appendln(uniteActorsAndResultsLinear(failure.scenario.initExecution, failure.results.initResults))
+        appendln(
+            uniteActorsAndResultsLinear(
+                failure.scenario.initExecution,
+                failure.results.initResults,
+                exceptionsDescription
+            )
+        )
     }
     if (failure.results.afterInitStateRepresentation != null)
         appendln("STATE: ${failure.results.afterInitStateRepresentation}")
     appendln("Parallel part:")
-    val parallelExecutionData = uniteParallelActorsAndResults(failure.scenario.parallelExecution, failure.results.parallelResultsWithClock)
+    val parallelExecutionData =
+        uniteParallelActorsAndResults(
+            failure.scenario.parallelExecution,
+            failure.results.parallelResultsWithClock,
+            exceptionsDescription
+        )
     appendln(printInColumns(parallelExecutionData))
     if (failure.results.afterParallelStateRepresentation != null) {
         appendln("STATE: ${failure.results.afterParallelStateRepresentation}")
     }
     if (failure.scenario.postExecution.isNotEmpty()) {
         appendln("Post part:")
-        appendln(uniteActorsAndResultsLinear(failure.scenario.postExecution, failure.results.postResults))
+        appendln(
+            uniteActorsAndResultsLinear(
+                failure.scenario.postExecution,
+                failure.results.postResults,
+                exceptionsDescription
+            )
+        )
     }
     if (failure.results.afterPostStateRepresentation != null && failure.scenario.postExecution.isNotEmpty()) {
         appendln("STATE: ${failure.results.afterPostStateRepresentation}")
     }
-    if (failure.results.parallelResultsWithClock.flatten().any { !it.clockOnStart.empty })
-        appendln("\n---\nvalues in \"[..]\" brackets indicate the number of completed operations \n" +
-            "in each of the parallel threads seen at the beginning of the current operation\n---")
+
+    val hints = mutableListOf<String>()
+    if (failure.results.parallelResultsWithClock.flatten().any { !it.clockOnStart.empty }) {
+        hints.add(
+            """
+                values in "[..]" brackets indicate the number of completed operations 
```

Comment:
According to this description, we need to replace "-" in clocks with "0" for better clarity.